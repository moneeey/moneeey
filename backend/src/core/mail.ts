import nodemailer from "nodemailer";
import fetch from "node-fetch";
import Mail from "nodemailer/lib/mailer";

import { SENDGRID_APIKEY, SMTP_URL } from "./config";

type MailOptions = {
  from: string;
  to: string;
  subject: string;
  text: string;
  html?: string;
};

const smtp_transport = () => nodemailer.createTransport(SMTP_URL);
const smtp_send = async (options: Mail.Options) => {
  const res = await smtp_transport().sendMail(options);
  if (res.accepted === options.to) {
    return { success: true };
  } else {
    return { success: false, error: res.response };
  }
};

const sendgrid_send = async ({
  from,
  to,
  subject,
  text,
  html,
}: MailOptions) => {
  const res = await fetch("https://api.sendgrid.com/v3/mail/send", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${SENDGRID_APIKEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      personalizations: [
        {
          to: [
            {
              email: to,
            },
          ],
        },
      ],
      from: {
        email: from,
      },
      subject: subject,
      content: [
        {
          type: "text/plain",
          value: text,
        },
        {
          type: "text/html",
          value: html,
        },
      ],
    }),
  });
  if (res.status === 202) {
    const ok = await res.text();
    return { success: true };
  } else {
    const error = await res.text();
    return { success: false, error };
  }
};

export const mail_send = ({
  from,
  to,
  text,
  html,
  subject,
}: MailOptions): Promise<{ success: boolean; error?: string }> => {
  if (SENDGRID_APIKEY) {
    return sendgrid_send({ from, to, html, text, subject });
  } else {
    return smtp_send({
      subject,
      from,
      to,
      html: html || text,
    });
  }
};

export type mail_send_fn = typeof mail_send;
