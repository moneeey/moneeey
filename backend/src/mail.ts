import { SENDGRID_API_KEY } from "./config.ts";
import { Logger } from "./logger.ts";

const logger = Logger("mail");

type MailOptions = {
	from: string;
	to: string;
	subject: string;
	text?: string;
	html?: string;
};

export const mailInternals = {
	apiKey: SENDGRID_API_KEY,
	fetch: (url: string, options: RequestInit) => {
		if (!mailInternals.apiKey || mailInternals.apiKey === "off") {
			logger.info("fetch", { url, options });
			return Promise.resolve(new Response("local sent", { status: 202 }));
		}
		return fetch(url, options);
	},
};

export const sendEmail = async ({
	from,
	to,
	subject,
	text,
	html,
}: MailOptions) => {
	const res = await mailInternals.fetch(
		"https://api.sendgrid.com/v3/mail/send",
		{
			method: "POST",
			headers: {
				Authorization: `Bearer ${mailInternals.apiKey}`,
				"Content-Type": "application/json",
			},
			body: JSON.stringify({
				personalizations: [
					{
						to: [
							{
								email: to,
							},
						],
					},
				],
				from: {
					email: from,
				},
				subject: subject,
				content: [
					{
						type: "text/plain",
						value: text,
					},
					{
						type: "text/html",
						value: html,
					},
				],
			}),
		},
	);
	if (res.status === 202) {
		return { success: true };
	}
	const error = await res.text();
	Logger("mail").error("error", { error });
	return { success: false };
};
