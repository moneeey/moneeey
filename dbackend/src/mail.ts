import { SENDGRID_API_KEY } from "./config.ts";

type MailOptions = {
  from: string;
  to: string;
  subject: string;
  text?: string;
  html?: string;
};

export const sendEmail = async ({
  from,
  to,
  subject,
  text,
  html,
}: MailOptions) => {
  if (!SENDGRID_API_KEY || SENDGRID_API_KEY === "off") {
    console.log(
      "sendEmail",
      JSON.stringify({ from, to, subject, text, html }, false, 4),
    );
    return { success: true };
  }
  const res = await fetch("https://api.sendgrid.com/v3/mail/send", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${SENDGRID_API_KEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      personalizations: [
        {
          to: [
            {
              email: to,
            },
          ],
        },
      ],
      from: {
        email: from,
      },
      subject: subject,
      content: [
        {
          type: "text/plain",
          value: text,
        },
        {
          type: "text/html",
          value: html,
        },
      ],
    }),
  });
  if (res.status === 202) {
    return { success: true };
  } else {
    const error = await res.text();
    console.log("sendEmail error", { error });
    return { success: false };
  }
};
