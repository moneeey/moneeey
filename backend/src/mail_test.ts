import { mailInternals, sendEmail } from "./mail.ts";
import {
  assert,
  assertSnapshot,
  withSpying,
  withSpyingLogger,
} from "./test.ts";

Deno.test(async function sendEmailSuccess(test) {
  const stub = await withSpying({
    object: mailInternals,
    method: "fetch",
    action: async (stub) => {
      stub.resolves(new Response("ok", { status: 202 }));
      assert.assertEquals(
        await sendEmail({
          from: "from@email.com",
          to: "to@email.com",
          subject: "The Subect",
          html: "<p>html</p>",
          text: "the text",
        }),
        { success: true },
      );
    },
  });
  assertSnapshot(test, stub.args);
});

Deno.test(async function sendEmailFail() {
  await withSpyingLogger(async (logger) => {
    await withSpying({
      object: mailInternals,
      method: "fetch",
      action: async (stub) => {
        stub.resolves(new Response("internal error", { status: 500 }));
        assert.assertEquals(
          await sendEmail({
            from: "from@email.com",
            to: "to@email.com",
            subject: "The Subect",
            html: "<p>html</p>",
            text: "the text",
          }),
          {
            success: false,
          },
        );
      },
    });
    assert.assertEquals(
      logger.args,
      [
        ["error", "[mail] error", { error: "internal error" }],
      ],
    );
  });
});

Deno.test(async function sendEmailInternalsLogger() {
  await withSpyingLogger(async (logger) => {
    mailInternals.apiKey = "off";
    assert.assertEquals(
      await sendEmail({
        from: "from@email.com",
        to: "to@email.com",
        subject: "The Subect",
        html: "<p>html</p>",
        text: "the text",
      }),
      {
        success: true,
      },
    );
    assert.assertEquals(
      logger.args.map(([level, message]) => [level, message]),
      [["info", "[mail] fetch"]],
    );
  });
});

Deno.test(async function sendEmailInternalsFetch() {
  await withSpyingLogger(async (logger) => {
    mailInternals.apiKey = "FAKE-KEY";
    await withSpying({
      object: globalThis,
      method: "fetch",
      action: async (stub) => {
        stub.resolves(new Response("ok", { status: 202 }));
        assert.assertEquals(
          await sendEmail({
            from: "from@email.com",
            to: "to@email.com",
            subject: "The Subect",
            html: "<p>html</p>",
            text: "the text",
          }),
          {
            success: true,
          },
        );
        assert.assertEquals(stub.args.length, 1);
        assert.assertEquals(stub.args[0][1].headers, {
          "Content-Type": "application/json",
          Authorization: "Bearer FAKE-KEY",
        });
      },
    });
    assert.assertEquals(logger.args, []);
  });
});
