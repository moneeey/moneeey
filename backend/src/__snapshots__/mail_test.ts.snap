export const snapshot = {};

snapshot[`sendEmailSuccess 1`] = `
[
  [
    "https://api.sendgrid.com/v3/mail/send",
    {
      body: '{"personalizations":[{"to":[{"email":"to@email.com"}]}],"from":{"email":"from@email.com"},"subject":"The Subect","content":[{"type":"text/plain","value":"the text"},{"type":"text/html","value":"<p>html</p>"}]}',
      headers: {
        "Content-Type": "application/json",
        Authorization: "Bearer off",
      },
      method: "POST",
    },
  ],
]
`;
